spec/lib/whitehall_importer/migrate_assets_spec.rb

```diff
@@ -0,0 +1,143 @@
+# frozen_string_literal: true
+
+RSpec.describe WhitehallImporter::MigrateAssets do
+  describe ".call" do
+    let(:asset_id) { "847150" }
+    let(:original_asset_url) { "https://asset-manager.gov.uk/blah/#{asset_id}/asset" }
+    let(:asset) { build(:whitehall_migration_asset_import) }
+
+    before :each do
+      stub_any_asset_manager_call
+    end
+
+    it "should take a WhitehallMigration::DocumentImport record as an argument" do
+      expect { described_class.call(build(:whitehall_migration_document_import)) }.not_to raise_error
+    end
+
+    it "should mark each asset as processing before its final state" do
+      whitehall_import = build(:whitehall_migration_document_import, assets: [asset])
+      expect(asset).to receive(:update!).with(state: "processing").once.ordered
+      expect(asset).to receive(:update!).with(hash_including(:state)).once.ordered
+      described_class.call(whitehall_import)
+    end
+
+    it "should skip migrating any assets that have already been processed" do
+      asset = create(:whitehall_migration_asset_import, state: "removed")
+      whitehall_import = build(:whitehall_migration_document_import, assets: [asset])
+      expect(asset).not_to receive(:update!)
+      described_class.call(whitehall_import)
+    end
+
+    it "should log GdsApi errors and put into an aborted state" do
+      whitehall_import = build(:whitehall_migration_document_import, assets: [asset])
+      allow(asset).to receive(:whitehall_asset_id).and_raise(GdsApi::BaseError.new("Some problem with GDS API"))
+      described_class.call(whitehall_import)
+      expect(asset.state).to eq("migration_failed")
+      expect(asset.error_message).to include("Some problem with GDS API")
+    end
+
+    it "should log standard errors and put into an aborted state" do
+      whitehall_import = build(:whitehall_migration_document_import, assets: [asset])
+      allow(asset).to receive(:whitehall_asset_id).and_raise(StandardError.new("Some error"))
+      described_class.call(whitehall_import)
+      expect(asset.state).to eq("migration_failed")
+      expect(asset.error_message).to include("Some error")
+    end
```

Comment:
I'm not sure we need both of these tests. I can see in the code we catch both but since we don't treat `GdsApi::BaseError` any differently to standard error it seems unnecessary to catch GDS Api specially?

Reply:
I see it as more likely that a GdsApi::BaseError could be thrown rather than a StandardError, but wanted to catch all errors (just in case) so we can fail gracefully.
My assumption was that if GdsApi::BaseError is not handled, our migration would crash rather than delegate the handling to the StandardError handler, so I've explicitly handled both. In hindsight I'm wondering if this is the case, and whether only capturing `StandardError` would actually be enough to capture custom errors too? If so, I agree we should remove the GdsApi::BaseError handler.

Reply:
GdsApi::BaseError inherits from StandardError so if we catch StandardError it's the same as catching GdsApi::BaseError.

catching StandardError is basically saying catch all errors, which is something we probably want to do if we want to log that a problem occurred rather than just crash.

Reply:
Thanks for confirming. I'll simplify this commit 👍